apps/api/v1/lib/utils/retrieveScopedAccessibleUsers.ts

```diff
@@ -0,0 +1,123 @@
+import prisma from "@calcom/prisma";
+import { MembershipRole } from "@calcom/prisma/enums";
+
+import { ScopeOfAdmin } from "~/lib/utils/isAdmin";
+
+type ScopeOfAdminValue = (typeof ScopeOfAdmin)[keyof typeof ScopeOfAdmin];
+type AccessibleUsersType = {
+  memberUserIds: number[];
+  adminUserId: number;
+};
+
+const getAllMemberships = async (userIds: number[]) => {
+  return await prisma.membership.findMany({
+    where: { userId: { in: userIds } },
+    select: {
+      userId: true,
+      team: {
+        select: {
+          id: true,
+        },
+      },
+    },
+  });
+};
+
+const getAllAdminMemberships = async (userId: number) => {
+  return await prisma.membership.findMany({
+    where: {
+      userId: userId,
+      OR: [{ role: MembershipRole.OWNER }, { role: MembershipRole.ADMIN }],
+    },
+    select: {
+      team: {
+        select: {
+          id: true,
+          isOrganization: true,
+        },
+      },
+    },
+  });
+};
+
+const getTeamsInOrganization = async (adminOrganizationId: number) => {
+  return await prisma.team.findMany({
+    where: {
+      parentId: adminOrganizationId,
+    },
+    select: {
+      id: true,
+    },
+  });
+};
+
+const getAllUsersInTeams = async (teamIds: number[]) => {
+  return await prisma.membership.findMany({
+    where: {
+      teamId: {
+        in: teamIds,
+      },
+    },
+    select: {
+      userId: true,
+    },
+  });
+};
+
+export const getAccessibleUsers = async ({
+  memberUserIds,
+  adminUserId,
+}: AccessibleUsersType): Promise<number[]> => {
+  const userMemberships = await getAllMemberships(memberUserIds);
+  const adminMemberships = await getAllAdminMemberships(adminUserId);
+
+  const accessibleUserIds = new Set<number>();
+
+  const adminOrganizationId = adminMemberships.filter((m) => m.team.isOrganization).map((m) => m.team.id)[0];
+  if (adminOrganizationId) {
+    userMemberships
+      .filter((um) => um.team.id === adminOrganizationId)
+      .forEach((um) => accessibleUserIds.add(um.userId));
```

Comment:
NIT: I don't like chaining array methods together. If we use reduce here we can bring this from `O(2n)` down to `O(n)`

Reply:
Nice find. Resolved!